### PR TITLE
Refactor Sync Script

### PR DESCRIPTION
**Pull request by popcor255**
# Summary

As a developer, it is difficult to make changes to a script and ensure that nothing is broken without tests. Test are added in this patch with 80% code coverage. The script has also been refactored to increase readability. This refactor is not a 'pure' refactor. The sync did not work properly. There is a list of bugs that reported the script generating broken links, #160, #126, #158, #133, #134. This refactor also includes a patch that fixes these bugs. The patch redirects a user to github if there is a relative link to a file that does not exist. This is important because some links will reference a snippet of code, test, or example. These artifacts are not docs, they are not rendered properly on the site. It is better to redirect the user to the intended artifact if the file is not synced. This prevents random 404s, and preserves the original behavior of the sync script. There are other approaches to fix this bug. This includes, manually editing links, downloading the files that the broken link are referencing in the sync config file, or removing the broken links. This is the best approach because broken links are automatically fixed and do not require manual intervention. This approach causes a side effect of not requiring the config file to be up-to date because the script will redirect the user to github. We want to encourage frequent sync. So eventually, this script should become `tekton-ified` (tekton task) and placed into the tekton catalog repo. Sync needs to become a tekon task in order for every repo to sync docs on push event.

# Changes

* Updated Documentation
* Added Tests with 80% code coverage
* Redirect examples, code snippets to original GitHub repo
* Refactored for readability

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/website/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

File: sync/README.md [hunk 1]
```diff
@@ -10,7 +10,7 @@ Google Cloud Platform credentials, and execute the following command:
 pip3 install -r requirements.txt
 python3 sync.py
 ```
-
+E
```
Comment (jjasghar) [inline sync/README.md:13]: typo?

**Review by jjasghar — COMMENTED** [review]
any chance we can see a demo of this running on Wednesday?

File: sync/README.md (hunk 1 above)
Comment (popcor255) [reply sync/README.md]: yep 😅

**Comment by popcor255** [comment]
> any chance we can see a demo of this running on Wednesday?

sure! 😄

**Review by afrittoli — COMMENTED** [review]
Thank you so much for this, and for adding all the tests too!! 👍 

The approach looks good to me. 
The main concern I have is on loading all the files into memory rather than processing them directly through `fileinput`.  Also we should be using a context to ensure files are always closed.

Could you create an issue in the `plumbing` repo about running the unit tests as part of CI?

I tested adding a few versions of `pipeline` in my local configs, which works well.
My preference would be to have the latest release displayed by default, and having access to `master` in the list of releases, however the naming `previous releases` does not fit anymore:



This is something that we could change in future PRs.
"Archived versions" also does not seem like a good name - the link points to the list of all releases, which are definitely not archived. Besides, we do not have a definition of what "archiving a release" means for Tekton. "Other version" might be a more fitting name?

I noticed that the order in the list of files does not seem to be honoured.
I tried adding a new file to my local config, at the end of the list, but it showed up second in the list:

image

Again this is something we could fix in a different PR.

File: sync/test_sync.py [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+        """ Verify if a string is a reference. A reference is
+        defined as  a string where its first character is a hashtag """
+        self.assertEqual(False, is_ref(""))
+        self.assertEqual(True, is_ref("#footer"))
+        self.assertEqual(False, is_ref("www.google.com"))
+
+    def test_remove_ending_forward_slash(self):
+        """ Remove a slash if it is the last character in a string """
+        expected = remove_ending_forward_slash("www.google.com/")
+        self.assertEqual("www.google.com", expected)
```
Comment (afrittoli) [inline sync/test_sync.py:55]: Here and in all tests below there is a bit of naming mismatch.
`expected` should be the value that you define as what you expect the function to return, and `actual` is the value that is returned by the function. 

Example of failure:
```
Traceback (most recent call last):
  File "/System/Volumes/Data/go/src/github.com/tektoncd/website/sync/test_sync.py", line 163, in test_is_url
    self.assertEqual(False, expected)
AssertionError: False != True
```
This error message seems to suggest that `expected` is `True` but it's actually the other way around. 

NIT: usually `assertEqual` takes `(actual, expected)`

So this test for instance should be:
```
        expected = "www.google.com"
        self.assertEqual(remove_ending_forward_slash("www.google.com/"), expected)
```

**Comment by afrittoli** [comment]
> This is something that we could change in future PRs.
> "Archived versions" also does not seem like a good name - the link points to the list of all releases, which are definitely not archived. Besides, we do not have a definition of what "archiving a release" means for Tekton. "Other version" might be a more fitting name?

https://github.com/tektoncd/website/pull/180

**Comment by popcor255** [comment]
Github is being weird I have to reopen a new pr :\

**Comment by popcor255** [comment]
@afrittoli https://github.com/tektoncd/website/pull/181
I made the changes at this pr ^

**Comment by popcor255** [comment]
> Thank you so much for this, and for adding all the tests too!! +1
> 
> The approach looks good to me.
> The main concern I have is on loading all the files into memory rather than processing them directly through `fileinput`. Also we should be using a context to ensure files are always closed.
> 
> Could you create an issue in the `plumbing` repo about running the unit tests as part of CI?
> 
> I tested adding a few versions of `pipeline` in my local configs, which works well.
> My preference would be to have the latest release displayed by default, and having access to `master` in the list of releases, however the naming `previous releases` does not fit anymore:
> 
> 
> 
> This is something that we could change in future PRs.
> "Archived versions" also does not seem like a good name - the link points to the list of all releases, which are definitely not archived. Besides, we do not have a definition of what "archiving a release" means for Tekton. "Other version" might be a more fitting name?
> 
> I noticed that the order in the list of files does not seem to be honoured.
> I tried adding a new file to my local config, at the end of the list, but it showed up second in the list:
> 
> image
> 
> Again this is something we could fix in a different PR.

Github wont update this pr even though my branch is updated. So this conversation should move to #181 .
To address the versioning issues. I 100% agree and I would like to make those changes. However, I do not want to change too much because this is only supposed to be refactor. I am really excited to make changes to fix versioning as well as ordering. I did make changes based on your feedback. Thank you so much for reviewing this PR! :heart:

**Pull request closed without merging**